firebase-firestore/src/androidTest/java/com/google/firebase/firestore/remote/RemoteStoreTest.java

```diff
@@ -0,0 +1,120 @@
+// limitations under the License.
+
+package com.google.firebase.firestore.remote;
+
+import static com.google.firebase.firestore.testutil.IntegrationTestUtil.waitFor;
+
+import android.support.test.InstrumentationRegistry;
+import android.support.test.runner.AndroidJUnit4;
+import com.google.firebase.database.collection.ImmutableSortedSet;
+import com.google.firebase.firestore.auth.User;
+import com.google.firebase.firestore.core.OnlineState;
+import com.google.firebase.firestore.local.LocalStore;
+import com.google.firebase.firestore.local.MemoryPersistence;
+import com.google.firebase.firestore.local.Persistence;
+import com.google.firebase.firestore.model.DocumentKey;
+import com.google.firebase.firestore.model.mutation.MutationBatchResult;
+import com.google.firebase.firestore.testutil.IntegrationTestUtil;
+import com.google.firebase.firestore.util.AsyncQueue;
+import com.google.firebase.firestore.util.Consumer;
+import io.grpc.Status;
+import java.util.concurrent.Semaphore;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+@RunWith(AndroidJUnit4.class)
+public class RemoteStoreTest {
+  @Test
+  public void testRemoteStoreStreamStopsWhenNetworkUnreachable() {
+    AsyncQueue testQueue = new AsyncQueue();
+    Datastore datastore =
+        new Datastore(
+            IntegrationTestUtil.testEnvDatabaseInfo(),
+            testQueue,
+            null,
+            InstrumentationRegistry.getContext());
+    Semaphore networkChangeSemaphore = new Semaphore(0);
+    RemoteStore.RemoteStoreCallback callback =
+        new RemoteStore.RemoteStoreCallback() {
+          public void handleRemoteEvent(RemoteEvent remoteEvent) {}
+
+          public void handleRejectedListen(int targetId, Status error) {}
+
+          public void handleSuccessfulWrite(MutationBatchResult successfulWrite) {}
+
+          public void handleRejectedWrite(int batchId, Status error) {}
+
+          public void handleOnlineStateChange(OnlineState onlineState) {
+            networkChangeSemaphore.release();
+          }
+
+          public ImmutableSortedSet<DocumentKey> getRemoteKeysForTarget(int targetId) {
+            return null;
+          }
+        };
+
+    FakeConnectivityMonitor connectivityMonitor = new FakeConnectivityMonitor();
+    Persistence persistence = MemoryPersistence.createEagerGcMemoryPersistence();
+    persistence.start();
+    LocalStore localStore = new LocalStore(persistence, User.UNAUTHENTICATED);
+    RemoteStore remoteStore =
+        new RemoteStore(callback, localStore, datastore, testQueue, connectivityMonitor);
+
+    waitFor(testQueue.enqueue(() -> remoteStore.forceEnableNetwork()));
+    drain(testQueue);
+    networkChangeSemaphore.drainPermits();
+
+    waitFor(
+        testQueue.enqueue(
+            () -> {
+              connectivityMonitor.goOffline();
+            }));
+    waitFor(networkChangeSemaphore);
+    drain(testQueue);
+
+    waitFor(testQueue.enqueue(() -> remoteStore.forceEnableNetwork()));
```

Comment:
I wasn't super wild about this test before, and now I'm liking it even less. It doesn't *quite* test what we want (though actually testing that the network recovers quickly when coming out of airplane mode is a bit challenging since you can't programatically alter the airplane mode state in api17+.) I may have to rethink this a bit... But in the meantime, if anyone has a shorter-term idea for making this test better, let me know.